File: apps/web/src/lib/automation-builder.ts
import { apiRequest } from "@/lib/api";
import type { WorkflowDefinition } from "@/types/workflow";

export interface WebhookPlan {
  name: string;
  direction: string;
}

export interface AutomationPlan {
  name: string;
  description: string | null;
  summary: string;
  trigger_type: string;
  steps_summary: string[];
  workflow: WorkflowDefinition;
  webhook: WebhookPlan | null;
}

export interface AutomationDeployResult {
  workflow_id: string;
  workflow_name: string;
  webhook_id: string | null;
  webhook_url: string | null;
  message: string;
}

export function buildAutomation(token: string, prompt: string) {
  return apiRequest<{ plan: AutomationPlan }>("/api/v1/automation/build", {
    method: "POST",
    token,
    body: JSON.stringify({ prompt }),
  });
}

export function deployAutomation(
  token: string,
  plan: AutomationPlan,
  activate = true,
) {
  return apiRequest<AutomationDeployResult>("/api/v1/automation/deploy", {
    method: "POST",
    token,
    body: JSON.stringify({ plan, activate }),
  });
}

File: apps/api/app/schemas/automation_builder.py
import uuid
from typing import Any

from pydantic import BaseModel, Field

from app.schemas.workflow import WorkflowDefinition


class AutomationBuildRequest(BaseModel):
    prompt: str = Field(min_length=3, max_length=4000)


class WebhookPlan(BaseModel):
    name: str
    direction: str = "incoming"


class AutomationPlan(BaseModel):
    name: str
    description: str | None = None
    summary: str
    trigger_type: str
    steps_summary: list[str] = Field(default_factory=list)
    workflow: WorkflowDefinition
    webhook: WebhookPlan | None = None


class AutomationBuildResponse(BaseModel):
    plan: AutomationPlan


class AutomationDeployRequest(BaseModel):
    plan: AutomationPlan
    activate: bool = True


class AutomationDeployResponse(BaseModel):
    workflow_id: uuid.UUID
    workflow_name: str
    webhook_id: uuid.UUID | None = None
    webhook_url: str | None = None
    message: str
